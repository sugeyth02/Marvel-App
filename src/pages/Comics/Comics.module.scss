@use './../../variables' as variables;
.container__cards {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  gap: 20px;
}
.container__filters {
  display: flex;
  flex-wrap: wrap;
  gap: 20px;
  justify-content: flex-end;
  padding: 30px 0;

  label {
    font-family: variables.$font-stack;
    padding: 10px;
    font-size: 1rem;
    font-weight: 500;
  }
  select {
    padding: 4px;
    font-family: variables.$font-stack;
    font-size: 0.9rem;
  }
}

@media (max-width: 435px) {
  .container__cards {
    justify-content: center;
    gap: 8px;
  }
  .container__filters {
    justify-content: flex-start;
  }
}