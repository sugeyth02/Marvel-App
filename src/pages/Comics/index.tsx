import React, { useCallback, useEffect, useState } from 'react';
import { useSearchParams } from 'react-router-dom';
import Card from '../../components/Card';
import Pagination from '../../components/Pagination';
import { IComic } from '../../entities/IComic';
import { useAppSelector } from '../../hook/useAppSelector';
import { getComics } from '../../redux/reducers/hiddenItems/selectors';
import services from '../../services/service';
import styles from './Comics.module.scss';

export default function Comics() {
  const [page, setPage] = useState(1);
  const [countItems, setCountItems] = useState(0);
  const [comics, setComics] = useState<IComic[]>([]);
  const [search, setSearch] = useSearchParams();
  const [pageQuery, nameQuery, formatQuery] = [
    search.get('page'),
    search.get('name'),
    search.get('format'),
  ];
  const hiddenComics = useAppSelector((state) => getComics(state.hiddenItems));
  useEffect(() => {
    const getData = async () => {
      try {
        const response = await services.getComics({
          limit: 24,
          ...(pageQuery !== null && {
            offset: `${(Number(pageQuery) - 1) * 24}`,
          }),
          ...(nameQuery !== null && { titleStartsWith: `${nameQuery}` }),
          ...(formatQuery !== null && { format: formatQuery }),
        });
        setComics(response.result);
        setCountItems(response.count);
      } catch (err) {}
    };
    getData();
  }, [nameQuery, formatQuery, pageQuery]);

  const handleFormatSelected = useCallback(
    (value: string) => {
      setSearch({
        page: '1',
        format: `${value}`,
      });
    },
    [setSearch]
  );

  useEffect(() => {
    setPage(1);
  }, [nameQuery, formatQuery, setPage]);

  useEffect(() => {
    setSearch({
      page: `${page}`,
      ...(nameQuery !== null && { name: `${nameQuery}` }),
      ...(formatQuery !== null && { format: formatQuery }),
    });
  }, [page, nameQuery, formatQuery, setSearch]);

  return (
    <div className={styles.container}>
      <div className={styles.container__filters}>
        <div className={styles.filters__select}>
          <label htmlFor='format'>Format:</label>
          <select
            name='format'
            id='format'
            onChange={(e) => {
              handleFormatSelected(e.target.value);
            }}
          >
            <option value='comic'>Comic</option>
            <option value='magazine'>Magazine</option>
            <option value='trade paperback'>Trade paperback</option>
            <option value='hardcover'>Hardcover</option>
            <option value='digest'>Digest</option>
            <option value='graphic novel'>Graphic novel</option>
            <option value='digital comic'>Digital comic</option>
            <option value='infinite comic'>Infinite comic</option>
          </select>
        </div>
      </div>
      <div className={styles.container__cards}>
        {comics
          .filter(
            (nohidden) =>
              !hiddenComics.some((hidden) => hidden.id === nohidden.id)
          )
          .map((comic) => (
            <Card
              key={comic.id}
              page='comics'
              info={{
                img: comic.thumbnail.path,
                title: comic.title,
                id: comic.id,
              }}
            />
          ))}
      </div>
      <Pagination
        setPage={setPage}
        currentPage={page}
        totalItems={countItems}
        itemsperPage={24}
      />
    </div>
  );
}
