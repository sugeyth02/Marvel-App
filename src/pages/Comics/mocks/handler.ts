import { rest } from 'msw';

const baseURL = 'https://gateway.marvel.com/v1/public/';

export const handlers = [
  rest.get(`${baseURL}comics`, (req, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json({
        data: {
          count: 1560,
          results: [
            {
              id: 82967,
              title: 'Marvel Previews (2017)',
              thumbnail: {
                path: 'http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available',
              },
            },
            {
              id: 376,
              title: 'Ant-Man (2003) #3',
              thumbnail: {
                path: 'http://i.annihil.us/u/prod/marvel/i/mg/d/70/4bc69c7e9b9d7',
              },
            },
          ],
        },
      })
    );
  }),
];
