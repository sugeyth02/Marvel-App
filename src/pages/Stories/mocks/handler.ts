import { rest } from 'msw';

const baseURL = 'https://gateway.marvel.com/v1/public/';

export const handlers = [
  rest.get(`${baseURL}stories`, (req, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json({
        data: {
          count: 1560,
          results: [
            {
              id: 7,
              title: 'Investigating the murder of a teenage girl, Cage suddenly learns that a three-way gang war is under way for control of the turf',
            },
            {
              id: 40,
              title: 'Ever-expanding their ranks, the Children of the Atom combat the evils threatening both mutants and humans Ã‘ like the Brotherhood',
            },
          ],
        },
      })
    );
  }),

  rest.get(`${baseURL}characters`, (req, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json({
        data: {
          results: [
            {
              id: 1011334,
              name: '3-D Man',
            },
            {
              id: 1009174,
              name: 'Beak',
            },
          ],
        },
      })
    );
  }),
];
