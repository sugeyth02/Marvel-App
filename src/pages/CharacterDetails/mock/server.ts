import { setupServer } from 'msw/node';
import { handlers } from './handler';
import { handlers as characterHandler } from '../../Characters/mock/handler';

const server = setupServer(...handlers,...characterHandler);
beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());
export default server;
