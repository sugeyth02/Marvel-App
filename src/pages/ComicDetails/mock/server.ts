import { setupServer } from 'msw/node';
import { handlers } from './handler';
import { handlers as comicsHandler } from '../../Comics/mocks/handler';

const server = setupServer(...handlers, ...comicsHandler);
beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());
export default server;
