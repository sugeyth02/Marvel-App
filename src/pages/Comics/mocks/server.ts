import { setupServer } from 'msw/node';
import { handlers } from './handler';


const server = setupServer(...handlers);
beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());
export default server;
