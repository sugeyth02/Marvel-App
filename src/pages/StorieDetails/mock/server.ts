import { setupServer } from 'msw/node';
import { handlers } from './handler';
import { handlers as storiesHandler } from '../../Stories/mocks/handler';

const server = setupServer(...handlers, ...storiesHandler);
beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());
export default server;
