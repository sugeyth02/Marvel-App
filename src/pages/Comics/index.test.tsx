import { render, screen, waitFor } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { store } from './../../redux/store';
import { Provider } from 'react-redux';
import './mocks/server';
import Comics from '.';

describe('Testing Comics list view', () => {
  beforeEach(() => {
    jest.spyOn(console, 'warn').mockImplementation(() => {});
    jest.spyOn(console, 'error').mockImplementation(() => {});
  });

  test('should renders a list of comics initially', async () => {
    // Render
    const { container } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Comics />
        </BrowserRouter>
      </Provider>
    );

    // Assertions
    await waitFor(
      () => {
        expect(
          screen.queryByText('Marvel Previews (2017)')
        ).toBeInTheDocument();
        expect(screen.queryByText('Ant-Man (2003) #3')).toBeInTheDocument();
        expect(container).toMatchSnapshot();
      },
      { timeout: 3000 }
    );
  });
});
