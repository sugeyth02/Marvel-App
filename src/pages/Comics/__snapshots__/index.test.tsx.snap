// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Comics list view should renders a list of comics initially 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="container__filters"
    >
      <div
        class="filters__select"
      >
        <label
          for="format"
        >
          Format:
        </label>
        <select
          id="format"
          name="format"
        >
          <option
            value="comic"
          >
            Comic
          </option>
          <option
            value="magazine"
          >
            Magazine
          </option>
          <option
            value="trade paperback"
          >
            Trade paperback
          </option>
          <option
            value="hardcover"
          >
            Hardcover
          </option>
          <option
            value="digest"
          >
            Digest
          </option>
          <option
            value="graphic novel"
          >
            Graphic novel
          </option>
          <option
            value="digital comic"
          >
            Digital comic
          </option>
          <option
            value="infinite comic"
          >
            Infinite comic
          </option>
        </select>
      </div>
    </div>
    <div
      class="container__cards"
    >
      <div
        class="container"
      >
        <div
          class="container__image"
          style="background-image: url(http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available/portrait_fantastic.jpg);"
        />
        <div
          class="container__info"
        >
          <h1>
            Marvel Previews (2017)
          </h1>
          <div
            class="buttons"
          >
            <button
              class="container__hide"
            >
              <i
                class="fa-solid fa-eye-slash"
              />
            </button>
            <button
              class="container__bookMark "
              data-testid="bookmark"
            >
              <i
                class="fa-solid fa-bookmark"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        class="container"
      >
        <div
          class="container__image"
          style="background-image: url(http://i.annihil.us/u/prod/marvel/i/mg/d/70/4bc69c7e9b9d7/portrait_fantastic.jpg);"
        />
        <div
          class="container__info"
        >
          <h1>
            Ant-Man (2003) #3
          </h1>
          <div
            class="buttons"
          >
            <button
              class="container__hide"
            >
              <i
                class="fa-solid fa-eye-slash"
              />
            </button>
            <button
              class="container__bookMark "
              data-testid="bookmark"
            >
              <i
                class="fa-solid fa-bookmark"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="container"
    >
      <button
        class="container__button"
        disabled=""
      >
        <i
          class="fa-solid fa-angle-left"
        />
      </button>
      <div
        class="numbers"
      />
      <button
        class="container__button"
      >
        <i
          class="fa-solid fa-angle-right"
        />
      </button>
    </div>
  </div>
</div>
`;
